### ISSUE TITLE
OID data type not supported in DDLDatabase

### ISSUE THREAD

**Issue by yuecelm** [jOOQ/jOOQ#8999]
introduce a new property `mode` to enable the following compatibility features in H2: http://www.h2database.com/html/features.html#compatibility

**Comment by yuecelm** [comment]
I use currently the OID data type in my sql scripts, H2 will trigger an error with DDLDatabase. Appending `MODE=PostgreSQL` to the JDBC URL would solve the error.

**Comment by yuecelm** [comment]
@lukaseder I would prepare a PR for this, some thoughts from your side?

**Comment by knutwannheden** [comment]
Thank you for your request.

We will discuss this with @lukaseder once he returns from vacation (in two weeks) and provide some feedback after that.

**Comment by yuecelm** [comment]
I wrote an test example which should fail when `MODE=PostgreSQL` is not set and OID is used in sql. Strangewise this test is not failing in master, but it is failing with jooq 3.11.11. I was looking into git logs but I didnt the related commit.

**Comment by knutwannheden** [comment]
Yes, that sounds strange. The parser should fail on the `CREATE TABLE ... WITH OIDS` clause.

Please note that I am coincidentally working on `DDLDatabase` in the scope of #8986 and will as a result refactor things somewhat.

**Comment by yuecelm** [comment]
I tested with this example in jooq 3.11.11:
```
CREATE TABLE local_storage (
  id CHAR(40) PRIMARY KEY,
  content OID NOT NULL
);
```
not sure if H2 is supporting `WITH OIDS` even with `MODE=PostgreSQL` set...

**Comment by yuecelm** [comment]
another difference: in jooq 3.11.11 I was using H2 1.4.197, master jooq is using H2 1.4.199

**Comment by yuecelm** [comment]
my current workaround in jooq 3.11.11:
```
        <!-- workaround for missing MODE=PostgreSQL in DDLDatabase, may be no more necessary with never jooq versions -->
        <plugin>
          <groupId>org.apache.maven.plugins</groupId>
          <artifactId>maven-antrun-plugin</artifactId>
          <version>1.8</version>
          <executions>
            <execution>
              <id>run-before</id>
              <phase>initialize</phase>
              <goals>
                <goal>run</goal>
              </goals>
              <configuration>
                <target>
                  <exec executable="sed">
                    <arg value="-i"/>
                    <arg value="s|OID|BLOB|"/>
                    <arg value="src/main/resources/db/migration/V1__init_schema.sql"/>
                  </exec>
                </target>
              </configuration>
            </execution>
            <execution>
              <id>run-after</id>
              <phase>process-sources</phase>
              <goals>
                <goal>run</goal>
              </goals>
              <configuration>
                <target>
                  <exec executable="sed">
                    <arg value="-i"/>
                    <arg value="s|BLOB|OID|"/>
                    <arg value="src/main/resources/db/migration/V1__init_schema.sql"/>
                  </exec>
                </target>
              </configuration>
            </execution>
          </executions>
        </plugin>
```

**Comment by knutwannheden** [comment]
Right. The jOOQ SQL parser will just keep the `OID` type as is (it doesn't really care what the column type is, but will map those it doesn't recognize to an unknown `Object.class` type). So if the H2 database swallows the type, it will work.

Possibly it is the H2 version which is responsible for the discrepancy. You could force Maven to use version 1.4.199 with jOOQ 3.11.11 and see if that solves your problem.

**Comment by katzyn** [comment]
All versions of H2 on Maven Central, even the 1.0.20061217, support the `OID` data type. Compatibility mode isn't needed.

H2 accepts `WITH OIDS` since 1.4.194. Compatibility mode isn't needed too.

**Comment by knutwannheden** [comment]
@katzyn Thanks for clarifying!

**Comment by yuecelm** [comment]
here is the error I get in jooq 3.11.11:
```
[INFO] --- jooq-codegen-maven:3.11.11:generate (default) @ JGFieldDeviceReports ---
[INFO] No <inputCatalog/> was provided. Generating ALL available catalogs instead.
[INFO] License parameters       
[INFO] ----------------------------------------------------------
[INFO]   Thank you for using jOOQ and jOOQ's code generator
[INFO]                          
[INFO] Database parameters      
[INFO] ----------------------------------------------------------
[INFO] Reading from: src/main/resources/db/migration
[INFO] Reading from: src/main/resources/db/migration/V1__init_schema.sql [*]
[ERROR] An exception occurred while parsing script source : src/main/resources/db/migration/*. Please report this error to https://github.com/jOOQ/jOOQ/issues/new
org.jooq.impl.ParserException: Unknown data type: [71:11] ...al_storage (
  id CHAR(40) PRIMARY KEY,
  content [*]OID NOT NULL
);

    at org.jooq.impl.ParserContext.exception (ParserImpl.java:8511)
    at org.jooq.impl.ParserImpl.parseDataTypePrefix (ParserImpl.java:7260)
    at org.jooq.impl.ParserImpl.parseDataType (ParserImpl.java:7023)
    at org.jooq.impl.ParserImpl.parseCreateTable (ParserImpl.java:2296)
    at org.jooq.impl.ParserImpl.parseCreate (ParserImpl.java:1834)
    at org.jooq.impl.ParserImpl.parseQuery (ParserImpl.java:674)
    at org.jooq.impl.ParserImpl.parse (ParserImpl.java:445)
    at org.jooq.impl.ParserImpl.parse (ParserImpl.java:432)
    at org.jooq.meta.extensions.ddl.DDLDatabase.load (DDLDatabase.java:207)
    at org.jooq.meta.extensions.ddl.DDLDatabase.load (DDLDatabase.java:181)
    at org.jooq.meta.extensions.ddl.DDLDatabase.load (DDLDatabase.java:194)
    at org.jooq.meta.extensions.ddl.DDLDatabase.create0 (DDLDatabase.java:157)
    at org.jooq.meta.AbstractDatabase.create (AbstractDatabase.java:231)
    at org.jooq.meta.AbstractDatabase.create (AbstractDatabase.java:220)
    at org.jooq.meta.AbstractDatabase.getDialect (AbstractDatabase.java:202)
    at org.jooq.codegen.AbstractGenerator.logDatabaseParameters (AbstractGenerator.java:155)
    at org.jooq.codegen.JavaGenerator.generate (JavaGenerator.java:299)
    at org.jooq.codegen.GenerationTool.run (GenerationTool.java:752)
    at org.jooq.codegen.GenerationTool.generate (GenerationTool.java:222)
    at org.jooq.codegen.maven.Plugin.execute (Plugin.java:197)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:289)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:229)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:415)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:356)
```

**Comment by yuecelm** [comment]
when the compatibility mode will not help here and the error is solved anyways in master, I think we can close this issue. would only good to know which jooq commit was resolving this issue. Some infos when a new jooq release is planned?

**Comment by knutwannheden** [comment]
This issue was resolved as part of #7776 and with commit 85e2b1c6ff1fd72952b59d4f3ea9884a6f366c88. Thank you very much for your investigations!

I will thus close this issue now. If another use case pops up for using the compatibility mode, we can open a new issue for that again. But at the moment I am rather reluctant towards this.

The 3.12 version will almost certainly be released later this month. But no promises!

**Comment by lukaseder** [comment]
Thanks a lot for your various comments and tests.

> Appending MODE=PostgreSQL to the JDBC URL would solve this error.

Numerous people have requested this kind of feature in the past. I have rejected it many times because the fact that we're using H2 behind the scenes should remain an implementation detail. We would not want to expose this detail through the `DDLDatabase` feature, and maintain and document it. I might change my opinion on this eventually, but currently, I think leaving this as internal saves everyone a lot of trouble.

The fact that we're using H2 was a quick win to offer something that works quite well on an 80/20 basis. In the future, we'll implement a much better `DDLDatabase`, where we'll simulate all the parsed SQL ourselves, rather than translating it to H2 (or any other random in-memory DB). This way, we'll be able to support a much bigger set of DDL. That will be done through https://github.com/jOOQ/jOOQ/issues/7034

A much quicker workaround for vendor specific DDL syntax that is not available in the jOOQ parser or in H2 would be to use the new parser ignore comment syntax, starting from jOOQ 3.12: https://github.com/jOOQ/jOOQ/issues/8325

Of course, replacing contents in your SQL statements before passing them to `DDLDatabase` is another viable option.
